workaround for broken wineg++

jasonMisaq opened the pull request:
Only changes paths to how they are addressed in Fedora

tresf commented:
@jasonMisaq thanks!

Since the original work-around is still perfectly valid for Ubuntu, can you please keep it in there?  Also, the comment above it should be simplified to simply state `# workaround for broken wineg++` without the specifics about Ubuntu.

tresf commented:
@jasonMisaq thanks!  Can you double-check the whitespace formatting?  It's a bit off now.  If you can't see it with your editor, take a look at it in the files: https://github.com/LMMS/lmms/pull/3130/files

jasonMisaq commented:
@tresf does this create syntax error? I am not much familiar with cmake files syntax

tresf commented on plugins/vst_base/CMakeLists.txt:
```diff
@@ -36,8 +36,12 @@ IF(LMMS_HOST_X86_64)
+	ENDIF()
+	#The following check works on Fedora systems 
+	IF("${WINEBUILD_OUTPUT}" MATCHES ".*lib64/wine/libwinecrt0.a.*")
+        SET(EXTRA_FLAGS ${EXTRA_FLAGS} -nodefaultlibs /usr/lib/i386/wine/libwinecrt0.a -luser32 -lkernel32 -lgdi32)
```
@jasonMisaq this looks like spaces instead of a tab, which is causing the extra indent.

tresf commented:
> @tresf does this whitespace formatting create a syntax error? I am not much familiar with cmake files syntax

@jasonMisaq no it does not, it's purely cosmetic but part of our code quality standards.

We haven't converted all of our source to use the same exact format, but we are pretty strict about matching a file to what it had before for readability purposes.

Once you fix that indent, we'll merge.  Thanks again!

jasp00 commented:
I am not sure this workaround is necessary. According to https://github.com/LMMS/lmms/issues/3129#issuecomment-261873431, 32-bit Wine was not installed in the first place.

jasonMisaq commented:
I do not really know @jasp00, but it solved https://github.com/LMMS/lmms/issues/3129 though. According to [Wine](https://wiki.winehq.org/FAQ#How_do_I_create_a_32_bit_wineprefix_on_a_64_bit_system.3F), there are significant bugs with 64 bit systems. So, I guess that's why installing 32-bit versions (i686) solved the build issue.

tresf commented:
> According to https://github.com/LMMS/lmms/issues/3129#issuecomment-261873431, 32-bit Wine was not installed in the first place.

Yes, but according to #3129 the build errors are confusing and obscure.  Won't this yield more sane error messages, such as `/usr/lib/i386-linux-gnu/wine/libwinecrt0.a not found`?  This seems much more sane over the current ` relocations from format elf64-x86-64 to format elf32-i386 is not supported`, no?

jasonMisaq commented:
@tresf, you are right, I was able to confirm this by removing wine-devel.i686 and wine.i686. Then, I built again by running `make`, and got this error:  
```
winegcc: File does not exist: /usr/lib/wine/libwinecrt0.a
plugins/vst_base/CMakeFiles/vstbase.dir/build.make:61: recipe for target 'plugins/RemoteVstPlugin' failed
make[2]: *** [plugins/RemoteVstPlugin] Error 2
CMakeFiles/Makefile2:8768: recipe for target 'plugins/vst_base/CMakeFiles/vstbase.dir/all' failed
make[1]: *** [plugins/vst_base/CMakeFiles/vstbase.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```
But, note that this is true considering the fact that we applied the workaround. Maybe if we did not, we would get the error we got in the first place (`relocations from format elf64-x86-64....`)

tresf commented:
> and got this error

This seems much more sane of an error.  Can you also test @jasp00's theory that reverting the patch has no effect on the end-result when the proper dependencies are met?  That way we at least know.

jasonMisaq commented:
Yes, @tresf I installed wine-devel.i686 again, reverted the patch. Removed CMakeCache.txt, did a `make clean`, and then ran `make` again. This time it built fine. I do not understand why did the error came up in the first place. So, maybe @jasp00 is right, the Fedora part (or even the Ubuntu part?) of the workaround might not be required. 

Now my ./plugins/vst_base/CMakeLists.txt looks like this:  
```
# workaround for broken wineg++ in WINE 1.4 (shipped e.g. with Ubuntu Precise)
        EXEC_PROGRAM( wineg++ ARGS "-v -m32 /dev/zero" OUTPUT_VARIABLE WINEBUILD_OUTPUT)
        IF("${WINEBUILD_OUTPUT}" MATCHES ".*x86_64-linux-gnu/wine/libwinecrt0.a.*")
                SET(EXTRA_FLAGS ${EXTRA_FLAGS} -nodefaultlibs /usr/lib/i386-linux-gnu/wine/libwinecrt0.a -luser32 -lkernel32 -lgdi32)
        ENDIF()
        #The following check works on Fedora systems
        #IF("${WINEBUILD_OUTPUT}" MATCHES ".*lib64/wine/libwinecrt0.a.*")
        #       SET(EXTRA_FLAGS ${EXTRA_FLAGS} -nodefaultlibs /usr/lib/i386/wine/libwinecrt0.a -luser32 -lkernel32 -lgdi32)
        #ENDIF()

```

tresf commented:
>  Fedora part (or even the Ubuntu part?) of the workaround might not be required.

Correct.  Since it yields a sane error message, I don't think we need to remove it.  Ideally we'd detect the environment properly never get into this logic, but that is for another day. 👍
